### ISSUE TITLE
Slow rendering on Pixel 8 device - OpenGL

### ISSUE THREAD

**Issue by Redsandro** [osmandapp/OsmAnd#18867]
### Description

When acquiring a replacement for my damaged 2019 smartphone, I made the mistake of assuming that Google's Pixel device would be the most hassle-free device because Android is also a Google product.

It turns out OsmAnd map drawing is very slow on this 2023 device. OpenGL rendering is slower than render engine version 1.

This issue doesn't seem to be tracked yet so I thought I'd open one.

### Steps to reproduce

Run OsmAnd render engine 2 (OpenGL) on a Pixel 8 device and compare with 5 year older device. 

### Actual result

Notice map drawing speed is considerably slower.

### Expected result

Map drawing speed equal or faster.

### Your Environment (required)

**WARNING** Crash-Logs **MAY** contain information you deem sensitive.
Review this **CAREFULLY** before posting your issue!

```
OsmAnd Version: 4.5.10
Android/iOS version: GrapheneOS Android 14
Device model: Google Pixel 8
```

**Comment by vshcherb** [comment]
That's very strange, we have Pixel 6a and it's very fast. Please try release version and share how the rendering looks like.

**Comment by Redsandro** [comment]
Okay. I just tried tried OsmAnd 4.6.11 released: 2023-12-28, it's the latest I can find. Rendering looks the same as the 4.5 version. It takes over 5 - 10 seconds to draw a screen in a big city.

**Comment by Redsandro** [comment]
Here is a screen recording using OsmAnd 4.6.11 on a Google Pixel 8.

https://github.com/osmandapp/OsmAnd/assets/1702193/83db78fc-87f6-4518-9ef5-3c8470da9045

Depending on zoom level, render time averages between 2 and 12 seconds.

I know OsmAnd draws a lot of detail, so it's hard to compare. Organic Maps draws screens in the same area mostly under a second.

**Comment by vshcherb** [comment]
This definitely not normal even for Pixel 6, but for Pixel 8 it's totally unexpected. We need to understand which optimizations are turned off or what's the issue.

**Comment by yuriiurshuliak** [comment]
Could you confirm if you're using the release version downloaded from the Google Play Store? 
Additionally, have you tested this on both the stock Android and GrapheneOS?

**Comment by Redsandro** [comment]
> Could you confirm if you're using the release version downloaded from the Google Play Store?

Yes. The recording is made with OsmAnd from on the Google Play store.

> have you tested this on both the stock Android and GrapheneOS?

Unfortunately I did not test this on the stock Android, because I don't allow [spyware](https://www.scss.tcd.ie/doug.leith/apple_google.pdf) on my phone. It would be great if someone with stock Android on a Pixel 8 can verify this. 

I ran 3dmark OpenGL ES 3.1 test and it was smooth as expected, if that helps. Geekbench 6 outperforms stock ROM for Pixel 8 in Single-Core (1634), Multi-Core (4133) and GPU OpenCL (5099). I have no issues with Google Maps, Organic Maps, Flitsmeister or Waze, although those last ones are probably too simple to notice.

**Comment by vshcherb** [comment]
We will try to test it further

**Comment by vshcherb** [comment]
Get a physical device of Pixel 7 but didn't reproduce it yet, will look for Pixel 8. So might be related to Graphene OS 14 somehow, so I will check for users with it.


https://github.com/osmandapp/OsmAnd/assets/1042025/7b7cce57-a1e0-4da6-8fe6-e5b6b1df5002

**Comment by sadellie** [comment]
Tested on Pixel 8. No lag on fresh install from F-Droid. 

This is a Graphene OS issue. I have seen multiple issues on different repos where Graphene OS users complain about lags.

**Comment by Redsandro** [comment]
Thank you for testing, @sadellie. 

It may be too quick to dub this a GrapheneOS issue though. It could be, but it doesn't necessarily make logical sense from an empirical point of view. I have nearly 100 apps installed, and OsmAnd is the only app that is lagging, severely. If at least some of the other OpenGL apps were lagging as well, I would not hesitate to blame GrapheneOS.

What does OsmAnd's OpenGL implementation uniquely do that is problematic exclusively on GrapheneOS, exclusively with OsmAnd? How have other map apps, like Maps, Mapy and Organic Maps, mitigated this problem? Can OsmAnd apply the same magic? I don't understand enough about the technical logistics to formulate a proper problem description and hypothesis. 

If someone knows specific eccentric OpenGL features that OsmAnd may use and the aforementioned map apps do not, at least I'd have some hypothesis to bring to the table. Otherwise, GrapheneOS will probably close the issue stating it has nothing to do with GrapheneOS.

**Comment by sadellie** [comment]
> @sadellie Please stop spreading misinformation about GrapheneOS.

There is no misinformation. 

Same device, same app, but different OS

**Comment by Redsandro** [comment]
@thestinger Wow, this is a big difference! I had no idea, not in the slightest, that this could be related. It didn't occur to me to try this toggle for OpenGL performance. This is indeed much, _much_ faster!

Again, I don't understand the technical reasons for this big difference especially since I hardly notice an impact in other OpenGL apps. So I don't know how much OsmAnd can do to perform better, and at what cost. I don't understand the security implications of the compatibility mode either. But I trust OsmAnd so either way I'm happy.

Let's stay composed, I'm sure no malice was intended. :slightly_smiling_face:  Thanks for the suggestion!

**Comment by Atrate** [comment]
> Here is a screen recording using OsmAnd 4.6.11 on a Google Pixel 8.
> osmand-v4611-pixel8.mp4
> 
> Depending on zoom level, render time averages between 2 and 12 seconds.
> 
> I know OsmAnd draws a lot of detail, so it's hard to compare. Organic Maps draws screens in the same area mostly under a second.

I can confirm that an extremely similar thing happens on the Fairphone 4 with DivestOS. Legacy is faster than OpenGL.

**Comment by Redsandro** [comment]
> I can confirm that an extremely similar thing happens on the Fairphone 4 with DivestOS. Legacy is faster than OpenGL.

@Atrate I don't know DivestOS, but for GrapheneOS the instructions from @thestinger fixed the problem. Perhaps DivestOS has similar protection modes that can be switched off for OsmAnd.

> (...) you should try enabling exploit protection compatibility mode. There's also an individual toggle for hardened_malloc to only disable hardened_malloc while leaving other features like extended address space enabled.

As for this issue and whether or not it should remain open: I don't know what OsmAnd does differently from other map apps, but my hypothesis is as follows:

1. It can be fixed;
2. Fixing it incurs a volunteer development cost;
3. Custom ROMs are a niche;
4. Disabling strict memory protection circumvents the problem.

I think the problem should be understood and documented, perhaps added to some kind of Kanban or [Wish list](https://en.wikipedia.org/wiki/MoSCoW_method#Prioritization_of_requirements) if it exists, and/or labeled _WontFix_ and closed.

**Comment by Atrate** [comment]
As far as I can see, there are no per-app toggles for exploit protection compat/hardened malloc in DivestOS. There's only a global "Enable secure app spawning" toggle. Maybe @SkewedZeppelin can shine some more light on the topic.

**Comment by Piscium** [comment]
Like the OP I have a Pixel 8 running GrapheneOS. Thestinger's suggestion of enabling exploit protection compatibility mode fixed it, so to speak (it made it fast).

I will make an extra comment for those trying to reproduce this issue that (like me!) are not Android experts. After toggling the relevant switch on GrapheneOS, force stop OsmAnd, and call it again (just swiping OsmAnd away is not enough, it seems OsmAnd still runs in the background).

**Comment by esatravi** [comment]
I run OsmAnd~ (v. 4.8.5) on Pixel 8 with the latest GrapheneOS (android 14)

Unfortunately no positive effect or improvement has been observed by toggling "exploit protection compatibility mode" or any other changes to security settings. 

OsmAnd map rendering is painfully slow (completely unusable actually) when Version 2 (OpenGL) rendering engine is selected. No workaround worked for me. 

Version 1 engine rendering speed is passable but still relatively slow. 

This is certainly not a hardware/CPU issue as Pixel 8 is a new and pretty strong mobile. The best benchmark for me is Organic Maps - it is literary *blazing* fast and has got features such as 3D buildings. I run it on my Pixel 8 and no delays with maps rendering is observed whatsoever. 

I suspect OsmAnd rendering engine code revision is required in order to remove processing bottlenecks.

**Comment by DmitryAlexei** [comment]
@esatravi Please provide a short video of your issue and coordinates of that spot

**Comment by Piscium** [comment]
> I run OsmAnd~ (v. 4.8.5) on Pixel 8 with the latest GrapheneOS (android 14)

OsmAnd~ on F-Droid is still on v. 4.8.4, where did you get v. 4.8.5?

> Unfortunately no positive effect or improvement has been observed by toggling "exploit protection compatibility mode" or any other changes to security settings.
> 
> OsmAnd map rendering is painfully slow (completely unusable actually) when Version 2 (OpenGL) rendering engine is selected. No workaround worked for me.
> 
> Version 1 engine rendering speed is passable but still relatively slow.
> 
> This is certainly not a hardware/CPU issue as Pixel 8 is a new and pretty strong mobile. The best benchmark for me is Organic Maps - it is literary _blazing_ fast and has got features such as 3D buildings. I run it on my Pixel 8 and no delays with maps rendering is observed whatsoever.
> 
> I suspect OsmAnd rendering engine code revision is required in order to remove processing bottlenecks.

I am also on Pixel 8 with the latest GrapheneOS (android 14), on v. 4.8.4, and cannot reproduce what you have. Like with "exploit protection compatibility mode" enabled it seems the same speed with version 1 or 2.

With "exploit protection compatibility mode" disabled it is crashing for me at the start, apparently due to memory tagging issue..

**Comment by esatravi** [comment]
> @esatravi Please provide a short video of your issue and coordinates of that spot

Here is the video of the slow map rendering in OsmAnd, when rendering engine is set to Version 2 (OpenGL)
https://github.com/user-attachments/assets/cad9d819-7c33-4562-9398-4636f5679d3a

Here, for comparison, rendering speed of Organic Maps, very responsive even with quick zooming in / out

https://github.com/user-attachments/assets/509b6a02-14f4-4175-9ee9-40d2a3bb68c0

**Comment by esatravi** [comment]
> It should perform as well as it does outside GrapheneOS with the exploit protection compatibility mode toggle enabled but it won't resolve any issue that's present elsewhere. It shouldn't really get much slower when using hardened_malloc though. We could provide more tuning for hardened_malloc performance vs. security settings but that might be an excessive amount of configuration...

Thank you for your comments. 
Unfortunately the experience is slow OpenGL rendering continues across various Pixel phones and GrapheneOS on them - I tried it on Pixel 3, Pixel 5, Pixel 8. OS doesn't get any modifications or google-play installation. 

My assumption was that since GrapheneOS is pure/degoogled android it should perform as well as Pixel stock OS. But it doesn't, for reasons unknown to me. I like OsmAnd and use with the Version 1 rendering engine instead.

**Comment by esatravi** [comment]
> > I run OsmAnd~ (v. 4.8.5) on Pixel 8 with the latest GrapheneOS (android 14)
> 
> OsmAnd~ on F-Droid is still on v. 4.8.4, where did you get v. 4.8.5?
> 
> > Unfortunately no positive effect or improvement has been observed by toggling "exploit protection compatibility mode" or any other changes to security settings.
> > OsmAnd map rendering is painfully slow (completely unusable actually) when Version 2 (OpenGL) rendering engine is selected. No workaround worked for me.
> > Version 1 engine rendering speed is passable but still relatively slow.
> > This is certainly not a hardware/CPU issue as Pixel 8 is a new and pretty strong mobile. The best benchmark for me is Organic Maps - it is literary _blazing_ fast and has got features such as 3D buildings. I run it on my Pixel 8 and no delays with maps rendering is observed whatsoever.
> > I suspect OsmAnd rendering engine code revision is required in order to remove processing bottlenecks.
> 
> I am also on Pixel 8 with the latest GrapheneOS (android 14), on v. 4.8.4, and cannot reproduce what you have. With "exploit protection compatibility mode" enabled it seems the same speed with version 1 or 2.
> 
> With "exploit protection compatibility mode" disabled it is crashing for me at the start, apparently due to a memory tagging issue..


I get my OsmAnd~ from f-droid repository: f-droid.org/packages/net.osmand.plus - suing droid-ify app. 
It is currently on version 4.8.5 (as of 9/8/2024)

Toggling  "exploit protection compatibility mode"  on or off doesn't effectuate any performance change, or does it crushes the OsmAnd app. I tried the toggle globally (in Settings/Security) or at the app level in its properties. 

I run out of ideas of enabling faster OpenGL mode...

**Comment by Piscium** [comment]
> I get my OsmAnd~ from f-droid repository: f-droid.org/packages/net.osmand.plus - suing droid-ify app. It is currently on version 4.8.5 (as of 9/8/2024)

You are right, it is 4.8.5. because it is the “suggested version” it was not installed by the F-droid app and not shown in the F-droid website at the top (one needs to scroll down quite a bit to show the 4.8.5 version). It is the second time I trip on this.

> Toggling "exploit protection compatibility mode" on or off doesn't effectuate any performance change, or does it crushes the OsmAnd app. I tried the toggle globally (in Settings/Security) or at the app level in its properties.

I upgraded to 4.8.5 but also with it it crashes on me due to a memory tagging error if I, while showing the map, with the finger change the centre of the map. It only takes a few seconds of doing so for it to crash.

A few months ago when I first installed osmand on the pixel 8 it also crashed, but only once at startup, due to “use after free”. That crash was not repeatable, unlike the current one.

There is (at least) one difference between your pixel 8 installation and mine: in mine i have installed Google Play. I don't know if that is the reason why we have different results doing the same tests.

A month ago or so I could install osmand from the Google Play Store in parallel with the version from F-droid. I can't anymore, no idea why. I wanted to test with the revision from  Google Play Store because it is newer (4.8.6). However, since I can't do a parallel install anymore, I am reluctant to test the Google Play version as that would mean uninstalling the F-droid version, and that would take time, to save my configuration from osmand, and reinstalling and restoring everything later.

> I run out of ideas how to to enable faster OpenGL mode...

Maybe temporarily install Google Play Services just for tests?

**Comment by spongebyte** [comment]
Pixel 9a user on GrapheneOS, OsmAnd 5.2.13 from Google Play. Same: extremely slow rendering, which gets a lot better with "Exploit protection compatibilitymode" toggled on.
